### PR TITLE
Support for input requests

### PR DESCRIPTION
**Pull request by SylvainCorlay**
input

### CODE REVIEW

File: src/xstream.cpp [hunk 1]
```diff
@@ -34,9 +34,28 @@ namespace xpyt
...
     PYBIND11_EMBEDDED_MODULE(xeus_python_stream, m) {
         py::class_<xstream>(m, "XPythonStream")
-            .def(py::init<std::string>())
-            .def("write", &xstream::write);
+            .def(py::init<std::string>());
+
+        m.def("write", write)
+         .def("input", input, py::arg("prompt") = "")
+         .def("raw_input", raw_input, py::arg("prompt") = "")
+         .def("getpass", getpass, py::arg("prompt") = "");
```
Comment (SylvainCorlay) [inline src/xstream.cpp]: Hesitant to put these into a `xeus_python_input` module.

Comment (SylvainCorlay) [reply src/xstream.cpp]: need to add an unused `stream` argument to `getpass`.

File: src/xinterpreter.cpp [hunk 2]
```diff
@@ -106,6 +107,8 @@ namespace xpyt
             return kernel_res;
         }
 
+        forward_input(allow_stdin);
```
Comment (SylvainCorlay) [inline src/xinterpreter.cpp]: I would like to make this a scope guard. For now, mimicking `ipykernel`.

File: src/xinterpreter.cpp [hunk 3]
```diff
@@ -251,4 +255,43 @@ namespace xpyt
...
+            builtins.attr("input") = xeus_python_stream.attr("input");
+        }
+        py::module getpass = py::module::import("getpass");
+        m_sys_getpass = getpass.attr("getpass");
+        getpass.attr("getpass") = xeus_python_stream.attr("getpass");
+    }
+
+    void interpreter::restore_input()
+    {
+        if (PY_MAJOR_VERSION == 3)
```
Comment (SylvainCorlay) [inline src/xinterpreter.cpp]: Could be a `#if` instead of `if`...

File: src/xstream.cpp [hunk 4]
```diff
@@ -34,9 +34,28 @@ namespace xpyt
...
+    }
+
+    void raw_input(const std::string& prompt)
+    {
+        xeus::get_interpreter().input_request(prompt, false);
+    }
+
+    void getpass(const std::string& prompt)
+    {
+        xeus::get_interpreter().input_request(prompt, true);
```
Comment (SylvainCorlay) [inline src/xstream.cpp]: TODO: Await response and flush stdin/stdout buffers?

https://github.com/ipython/ipykernel/blob/118b684f75d51d72819ca90b69a245e9b02067a5/ipykernel/kernelbase.py#L875

File: src/xinterpreter.cpp (hunk 2 above)
Comment (martinRenou) [inline src/xinterpreter.cpp]: You should maybe call `forward_input` only if `allow_stdin` is true? And remove the argument for the `forward_input` method, as you don't use it.

File: src/xinterpreter.cpp [hunk 5]
```diff
@@ -162,6 +165,7 @@ namespace xpyt
             kernel_res["evalue"] = evalue;
             kernel_res["traceback"] = traceback;
         }
+        restore_input();
```
Comment (martinRenou) [inline src/xinterpreter.cpp]: Same for `restore_input`?

Comment (SylvainCorlay) [reply src/xinterpreter.cpp]: Resolved, now using a scope guard.

File: src/xinterpreter.cpp (hunk 2 above)
Comment (SylvainCorlay) [reply src/xinterpreter.cpp]: Resolved as implemented in a scope guard.

File: src/xstream.cpp (hunk 4 above)
Comment (SylvainCorlay) [reply src/xstream.cpp]: Closing and adding a comment in the thread.

File: src/xinterpreter.cpp (hunk 2 above)
Comment (SylvainCorlay) [reply src/xinterpreter.cpp]: Resolved.

File: src/xstream.cpp (hunk 1 above)
Comment (SylvainCorlay) [reply src/xstream.cpp]: Done.

**Comment by SylvainCorlay** [comment]
TODO: Await response and flush stdin/stdout buffers.

https://github.com/ipython/ipykernel/blob/118b684f75d51d72819ca90b69a245e9b02067a5/ipykernel/kernelbase.py#L875

**Comment by SylvainCorlay** [comment]
Merging! cc @jcfr @ihnorton (cf screencast above).

File: src/xinput.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        py::module xeus_python_input = py::module::import("xeus_python_input");
+
+        // Forward input()
+        py::module builtins = py::module::import(XPYT_STRINGIFY(XEUS_PYTHON_BUILTINS));
+        m_sys_input = builtins.attr("input");
+        builtins.attr("input") = allow_stdin ? xeus_python_input.attr("input")
+                                             : xeus_python_input.attr("notimplemented");
+
+        // Forward raw_input()
+        if (PY_MAJOR_VERSION == 2)
```
Comment (martinRenou) [inline src/xinput.cpp]: Could it be a `#if PY_MAJOR_VERSION <= 2`?

File: src/xinput.cpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    input_redirection::~input_redirection()
+    {
+        // Restore input()
+        py::module builtins = py::module::import(XPYT_STRINGIFY(XEUS_PYTHON_BUILTINS));
+        builtins.attr("input") = m_sys_input;
+
+        // Restore raw_input()
+        if (PY_MAJOR_VERSION == 2)
```
Comment (martinRenou) [inline src/xinput.cpp]: `#if PY_MAJOR_VERSION <= 2`?

Comment (SylvainCorlay) [reply src/xinput.cpp]: I don't think we will ever support Python 1!

File: src/xinterpreter.cpp.orig [hunk 8]
```diff
@@ -0,0 +1,392 @@
+/***************************************************************************
```
[the remark this answers is not in the document]
Comment (SylvainCorlay) [reply src/xinterpreter.cpp.orig]: Removing it on the master branch.

File: src/xcomm.cpp.orig [hunk 9]
```diff
@@ -0,0 +1,166 @@
+/***************************************************************************
```
[the remark this answers is not in the document]
Comment (SylvainCorlay) [reply src/xcomm.cpp.orig]: idem

File: src/xinput.cpp (hunk 7 above)
Comment (martinRenou) [reply src/xinput.cpp]: I'm not talking about Python1. What I meant is that we could check that at compilation time

Comment (SylvainCorlay) [reply src/xinput.cpp]: Yes, that is what I was doing earlier. Why not!

**Pull request merged**